File: Zds/Zds.Flow/Stopwatch/Stopwatch.vb
Namespace Stopwatch
    <DebuggerStepThrough>
    Public Class Stopwatch
        Implements IStopwatch
        Private _IsRunning As Boolean = False
        Private _Stopwatch As IStopwatch
        Private _Time As TimeSpan
        Public ReadOnly Property IsRunning As Boolean Implements Interfaces.IStartStopable.IsRunning
            Get
                Return _IsRunning
            End Get
        End Property
        Public Property Elapsed As TimeSpan Implements IStopwatch.Elapsed
            Get
                If _IsRunning = True Then
                    Return _Stopwatch.Elapsed - _Time
                Else
                    Return _Time
                End If
            End Get
            Set(value As TimeSpan)
                If _IsRunning = True Then
                    _Time = _Stopwatch.Elapsed - value
                Else
                    _Time = value
                End If
            End Set
        End Property
        Public Sub Start() Implements Interfaces.IStartStopable.Start
            If _IsRunning = False Then
                _Time = _Stopwatch.Elapsed - Me.Elapsed
                _IsRunning = True
            End If
        End Sub
        Public Sub [Stop]() Implements Interfaces.IStartStopable.Stop
            If _IsRunning = True Then
                _Time = Elapsed
                _IsRunning = False
            End If
        End Sub
        Public Sub Reset() Implements Interfaces.IResetable.Reset
            [Stop]()
            _Time = TimeSpan.Zero
        End Sub
        Sub New(Reference As IStopwatch)
            _Stopwatch = Reference
        End Sub
        Sub New()
            Me.New(UniversalStopwatch)
        End Sub
        Public Shared ReadOnly UniversalStopwatch As IStopwatch
        Shared Sub New()
            UniversalStopwatch = New SystemStopwatch
            UniversalStopwatch.Start()
        End Sub
    End Class
End Namespace

File: Flow/Flow/Stopwatch/Stopwatch.cs
using System;
using System.Collections.Generic;
using System.Text;
using Flow.Interfaces;

namespace Flow.Stopwatch
{
    public class Stopwatch : IStopwatch
    {
        private IStopwatch stopwatch = GlobalStopwatch;
        private TimeSpan time = TimeSpan.Zero;
        public bool IsRunning { get; private set; }
        public IStopwatch Reference
        {
            get => stopwatch;
            set
            {
                Stop();
                stopwatch = value;
                Start();
            }
        }
        public TimeSpan Elasped
        {
            get
            {
                if (IsRunning) return stopwatch.Elasped - time;
                else return time;
            }
            set
            {
                if (IsRunning) time = stopwatch.Elasped - value;
                else time = value;
            }
        }
        public void Start()
        {
            if (!IsRunning)
            {
                time = stopwatch.Elasped - time;
                IsRunning = true;
            }
        }
        public void Stop()
        {
            if (IsRunning)
            {
                time = Elasped;
                IsRunning = false;
            }
        }
        public void Reset()
        {
            Stop();
            time = TimeSpan.Zero;
        }
        public static readonly IStopwatch GlobalStopwatch = new SystemStopwatch();
        public Stopwatch() { }
        public Stopwatch(IStopwatch reference)
        {
            this.stopwatch = reference;
        }

        static Stopwatch() => GlobalStopwatch.Start();
    }

    public interface IStopwatch : IStartStopable
    {
        TimeSpan Elasped { get; set; }
    }
}
